synapse/handlers/message.py

```diff
@@ -1135,11 +1135,13 @@ async def create_new_client_event(
             )
             state_events = await self.store.get_events_as_list(state_event_ids)
             # Create a StateMap[str]
-            state_map = {(e.type, e.state_key): e.event_id for e in state_events}
```

Comment:
It looks like the `state_map` function parameter was added more recently in https://github.com/matrix-org/synapse/pull/13487 which caused the collision. Whereas the `state_map` variable was in place since https://github.com/matrix-org/synapse/pull/12083 and was only used as a temporary variable within itself at the time so this variable shadowing case seems legit.

(just pulled out of https://github.com/matrix-org/synapse/pull/12083)
https://github.com/matrix-org/synapse/blob/cd930564b8423a88725b233f7567e5bd7a63446b/synapse/handlers/message.py#L972-L976

For reference and confusingly, all of the `for_batch` and `state_map` parameter stuff isn't for MSC2716 `/batch_send` (it's all new from the room creation batch event creation). It really feels like https://github.com/matrix-org/synapse/pull/13487 added some duplication of existing logic that could be covered by `state_event_ids`/`auth_event_ids` but I haven't delved into all of that code.

---

> This could happen if `for_batch=True`, `state_event_ids` is not-`None`.
> 
> So I guess the question is -- is it valid input if `state_event_ids` and `state_map` to both be not-`None`?
> 
> I think this might be a question for @MadLittleMods 
>
> *-- @clokep, https://github.com/matrix-org/synapse/pull/14575#pullrequestreview-1196453788*

As mentioned, since none of the stuff I've worked on for MSC2716 uses these parameters, I don't think there is a case where they interact unless @H-Shay has refactored some of that code and is probably the one to better answer.

Reply:
Ah, my misunderstanding! Sorry about that. I assumed it all had to do with MSC2716 due to the "batch" comment in it. 😢

Reply:
> It really feels like https://github.com/matrix-org/synapse/pull/13487 added some duplication of existing logic that could be covered by `state_event_ids`/`auth_event_ids` but I haven't delved into all of that code.

@H-Shay Can you create a follow-up issue to track this?

And a separate one for "If we could somehow differentiate our internal batched event creating from [MSC2716](https://github.com/matrix-org/matrix-spec-proposals/pull/2716) that would be fabulous..." (https://github.com/matrix-org/synapse/pull/14575#pullrequestreview-1196533727)

Reply:
> Can you create a follow-up issue to track this?

Sure thing, I have created #14689 for the naming issue and https://github.com/matrix-org/synapse/issues/14688 for the possible logic duplication in `create_new_client_event`.